pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	maxBins        = 16
+	nnLowWatermark = 2 // the number of peers in consecutive deepest bins that constitute as nearest neighbours
+)
+
+var (
+	errMissingAddressBookEntry = errors.New("addressbook underlay entry not found")
+	timeToRetry                = 30 * time.Second
+)
+
+type binSaturationFunc func(bin, depth uint8, peers *pslice.PSlice) bool
+
+// Options for injecting services to Kademlia.
+type Options struct {
+	Base           swarm.Address
+	Discovery      discovery.Driver
+	AddressBook    addressbook.Interface
+	P2P            p2p.Service
+	SaturationFunc binSaturationFunc
+	Logger         logging.Logger
+}
+
+// Kad is the Swarm forwarding kademlia implementation.
+type Kad struct {
+	base           swarm.Address         // this node's overlay address
+	discovery      discovery.Driver      // the discovery driver
+	addressBook    addressbook.Interface // address book to get underlays
+	p2p            p2p.Service           // p2p service to connect to nodes with
+	saturationFunc binSaturationFunc     // pluggable saturation function
+	connectedPeers *pslice.PSlice        // a slice of peers sorted and indexed by po, indexes kept in `bins`
+	knownPeers     *pslice.PSlice        // both are po aware slice of addresses
+	depth          uint8                 // current neighborhood depth
+	depthMu        sync.RWMutex          // protect depth changes
+	manageC        chan struct{}         // trigger the manage forever loop to connect to new peers
+	waitNext       map[string]time.Time  // sancation connections to a peer, key is overlay string and value is time to next retry
+	waitNextMu     sync.Mutex            // synchronize map
+	logger         logging.Logger        // logger
+	quit           chan struct{}         // quit channel
+	done           chan struct{}         // signal that `manage` has quit
+}
+
+// New returns a new Kademlia.
+func New(o Options) *Kad {
+	if o.SaturationFunc == nil {
+		o.SaturationFunc = binSaturated
+	}
+
+	k := &Kad{
+		base:           o.Base,
+		discovery:      o.Discovery,
+		addressBook:    o.AddressBook,
+		p2p:            o.P2P,
+		saturationFunc: o.SaturationFunc,
+		connectedPeers: pslice.New(maxBins),
+		knownPeers:     pslice.New(maxBins),
+		manageC:        make(chan struct{}, 1),
+		waitNext:       make(map[string]time.Time),
+		logger:         o.Logger,
+		quit:           make(chan struct{}),
+		done:           make(chan struct{}),
+	}
+
+	go k.manage()
+	return k
+}
+
+// manage is a forever loop that manages the connection to new peers
+// once they get added or once others leave.
+func (k *Kad) manage() {
+	var peerToRemove swarm.Address
+	defer close(k.done)
+
+	for {
+		select {
+		case <-k.quit:
+			return
+		case <-k.manageC:
+			err := k.knownPeers.EachBinRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
```

Comment:
`manage` loop is not meant to be efficient right now and it was not intended to be. 

my personal opinion is that even if it takes your node a few seconds to connect to all relevant peers when you start it up - this is not a big price to pay at this moment. we can optimize later, but establishing connections is actually not _so_ critical for performance since it is not supposed to happen all that often anyway, and you'll have other nodes dialing in anyway.

peer selection for retrieve requests on the other hand should be very fast and efficient and i am sure that we will need to spend time optimizing for performance there, but i would not spend tons of engineering hours optimizing here at this point until it proves necessary.

Reply:
> my personal opinion is that even if it takes your node a few seconds to connect to all relevant peers when you start it up 

Cold start are a big disadvantages of bittorent. If not now, we should aim for faster saturation of our kademlia table.